chain/client/src/client.rs

```diff
@@ -220,18 +294,8 @@ impl Client {
 
         debug!(target: "client", "{:?} Producing block at height {}, parent {} @ {}", block_producer.account_id, next_height, prev.inner_lite.height, format_hash(head.last_block_hash));
 
-        if self.runtime_adapter.is_next_block_epoch_start(&head.last_block_hash)? {
-            if !self.chain.prev_block_is_caught_up(&prev_prev_hash, &prev_hash)? {
-                // Currently state for the chunks we are interested in this epoch
-                // are not yet caught up (e.g. still state syncing).
-                // We reschedule block production.
-                // Alex's comment:
-                // The previous block is not caught up for the next epoch relative to the previous
-                // block, which is the current epoch for this block, so this block cannot be applied
-                // at all yet, block production must to be rescheduled
-                debug!(target: "client", "Produce block: prev block is not caught up");
-                return Ok(None);
-            }
+        if let Some(res) = self.should_reschedule_block(&head, &prev_hash, &prev_prev_hash)? {
```

Comment:
1. `should_reschedule_block` should return `Result<bool>`, and return `Ok(None)` here if it returned true. It never actually returns any `Block`. Returning `bool` from a function that has a name `should_...` makes a lot of sense :)
2. Move the logic that compares heights and whether `me` is a block producer inside `should_reschedule_block`, so that there's only one function instead of three (nothing above really prevents rearranging the logic and making all three checks adjacent).

Reply:
> Move the logic that compares heights and whether me is a block producer inside should_reschedule_block, so that there's only one function instead of three (nothing above really prevents rearranging the logic and making all three checks adjacent).

1. It may be impossible because of the way how we use adversarial variables. They may turn on or not, and we should handle it properly.
2. It increase complexity of adversarial code.
3. It may be extremely hard and unclear how to add new things to adversarial block producing because of lots of returns and obfuscated conditions.

I suggest to keep several functions instead which clearly shows which conditions should be met by turning on exact adversarial controls.

Reply:
1. Done

2. I moved some logic inside `should_reschudule_block`, but couldn't get rid of the three functions, since they are designed in such a way that adversarial code is inside the feature and return fast when needed. Unwrapping them will require more complex code.

The main feature with this code is to have all the adversarial code exactly behind the feature, and try not to affect readability too much.